hedera-node/src/main/java/com/hedera/services/contracts/operation/HederaLogOperation.java

```diff
@@ -0,0 +1,69 @@
+package com.hedera.services.contracts.operation;
+
+import com.google.common.collect.ImmutableList;
+import com.hedera.services.store.contracts.HederaStackedWorldStateUpdater;
+import com.hedera.services.utils.EntityNum;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.tuweni.bytes.Bytes;
+import org.hyperledger.besu.datatypes.Address;
+import org.hyperledger.besu.evm.EVM;
+import org.hyperledger.besu.evm.Gas;
+import org.hyperledger.besu.evm.frame.ExceptionalHaltReason;
+import org.hyperledger.besu.evm.frame.MessageFrame;
+import org.hyperledger.besu.evm.gascalculator.GasCalculator;
+import org.hyperledger.besu.evm.log.Log;
+import org.hyperledger.besu.evm.log.LogTopic;
+import org.hyperledger.besu.evm.operation.AbstractOperation;
+
+import java.util.Optional;
+
+import static org.apache.tuweni.bytes.Bytes32.leftPad;
+import static org.hyperledger.besu.evm.internal.Words.clampedToLong;
+
+public class HederaLogOperation extends AbstractOperation {
+	private static final Logger log = LogManager.getLogger(HederaLogOperation.class);
+
+	private static final Address UNRESOLVABLE_ADDRESS_STANDIN = EntityNum.MISSING_NUM.toEvmAddress();
+
+	private final int numTopics;
+
+	public HederaLogOperation(final int numTopics, final GasCalculator gasCalculator) {
+		super(0xA0 + numTopics, "LOG" + numTopics, numTopics + 2, 0, 1, gasCalculator);
+		this.numTopics = numTopics;
+	}
+
+	@Override
+	public OperationResult execute(final MessageFrame frame, final EVM evm) {
+		final long dataLocation = clampedToLong(frame.popStackItem());
+		final long numBytes = clampedToLong(frame.popStackItem());
+
+		final Gas cost = gasCalculator().logOperationGasCost(frame, dataLocation, numBytes, numTopics);
+		final Optional<Gas> optionalCost = Optional.of(cost);
+		if (frame.isStatic()) {
+			return new OperationResult(
+					optionalCost, Optional.of(ExceptionalHaltReason.ILLEGAL_STATE_CHANGE));
+		} else if (frame.getRemainingGas().compareTo(cost) < 0) {
+			return new OperationResult(optionalCost, Optional.of(ExceptionalHaltReason.INSUFFICIENT_GAS));
+		}
+
+		final var addressOrAlias = frame.getRecipientAddress();
```

Comment:
If I understand the implication right, the address could be either the create2 address or the hedera address, depending on if the client called via the create2 or the hedera address. The goal is to then assure that all logs log to the same address based on calling convention?

I think we need to ratchet this up another level.  The recipient address (and thus result of the ADDRESS operation) always returns either the hedera version or the create 2 version, no matter how it is called.  This has contact security implication for when items like contract owner are recorder for access control.  If you call the contract the wrong way it doesn't have the expected permissions, and that could present problems.

My vote is that the receiver address is always the alias address for a contract if it has one in-effect at the time of the call, regardless of how it's called.  My reason is I expect that there will be contracts that may not check the return value of create2, presume the calculation is what they expect, and set ACLs based on the create2 address.

Reply:
> My vote is that the receiver address is always the alias address for a contract if it has one in-effect at the time of the call, regardless of how it's called. 

This is the current behavior, since we override `CallOperation#address()` and `StaticCallOperation#address()` [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/contracts/operation/HederaCallOperation.java#L69) and [here](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/contracts/operation/HederaCallOperation.java#L69), respectively.

The problem for specifically `ContractID`'s in a record's `ContractFunctionResult` is that, by the time the transaction finishes and we are converting `Log` objects to gRPC [in `TransactionProcessingResult`](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/contracts/execution/TransactionProcessingResult.java#L183), it may no longer be **possible** to resolve the aliases to a Hedera `0.0.X` id. (E.g. because the `CREATE2` contract self-destructed during the transaction.)

But Mirror Node team has requested that any `ContractID` in a record be in the `0.0.X` form. So we have to intervene here, to make sure the `Log` always contains a mirror address that can be parsed to a `0.0.X` id.

Reply:
Disagree and commit.  This will come back to bite us.  But we need immediate action for mirror node.